compiler/src/dmd/dtemplate.d

```diff
@@ -4141,185 +4199,192 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                     goto Lnomatch;
 
             L2:
-                for (size_t i = 0; 1; i++)
+                if (!resolveTemplateInstantiation(t.tempinst.tiargs, &t.tempinst.tdtypes, tempdecl, tp, dedtypes))
+                    goto Lnomatch;
+            }
+            visit(cast(Type)t);
+            return;
+
+        Lnomatch:
+            //printf("no match\n");
+            result = MATCH.nomatch;
+        }
+
+        private bool resolveTemplateInstantiation(Objects* tiargs, Objects* tdtypes, TemplateDeclaration tempdecl, TypeInstance tp, Objects* dedtypes)
```

Comment:
This function appears to be code cut&pasted out of the original. But it is very difficult to verify it. I suggest that the large refactoring go in a separate PR as a predecessor to this PR.

Reply:
You are right. I've opened a new PR: https://github.com/dlang/dmd/pull/15263